client/connection.cpp

```diff
@@ -92,41 +92,38 @@ bool Connection::send(const void* buf, size_t len) {
   }
 }
 
-bool Connection::receive(std::string& dest) {
+bool Connection::receive_(std::function<void()> assignToDestination) {
   clearError();
   try {
-    bool res = sock_->recv(&recvmsg_);
-    if (res) {
-      dest.assign(recvmsg_.data<char>(), recvmsg_.size());
+    bool response = sock_->recv(&recvmsg_);
+    if (response) {
+      assignToDestination();
     } else {
       errnum_ = EAGAIN;
       errmsg_ = ERRMSG_TIMEOUT_EXCEEDED;
     }
-    return res;
-  } catch (zmq::error_t& e) {
-    errnum_ = e.num();
-    errmsg_ = e.what();
+    return response;
+  } catch (zmq::error_t& exception) {
+    if (exception.num() == EINTR) {
+      return receive_(assignToDestination);
```

Comment:
I'd prefer a loop to be honest, you could just continue here since all other code paths will return.

Reply:
Why a loop? Modern C++ compilers have tail-call optimization.